pkg/streamingpromql/operators/and_binary_operation.go

```diff
@@ -0,0 +1,292 @@
+			// This is the last series for this group, return it to the pool.
+			thisSeriesGroup.Close(a.MemoryConsumptionTracker)
+		}
+
+		return filteredData, nil
+	}
+}
+
+// readRightSideUntilGroupComplete reads series from the right-hand side until all series for desiredGroup have been read.
+func (a *AndBinaryOperation) readRightSideUntilGroupComplete(ctx context.Context, desiredGroup *andGroup) error {
+	for a.nextRightSeriesIndex <= desiredGroup.lastRightSeriesIndex {
+		groupForRightSeries := a.rightSeriesGroups[0]
+		a.rightSeriesGroups = a.rightSeriesGroups[1:]
+
+		data, err := a.Right.NextSeries(ctx)
+		if err != nil {
+			return err
+		}
+
+		if groupForRightSeries != nil {
+			if err := groupForRightSeries.AccumulateRightSeriesPresence(data, a.MemoryConsumptionTracker, a.timeRange); err != nil {
+				return err
+			}
+		}
+
+		types.PutInstantVectorSeriesData(data, a.MemoryConsumptionTracker)
+		a.nextRightSeriesIndex++
+	}
+
+	return nil
+}
+
+func (a *AndBinaryOperation) ExpressionPosition() posrange.PositionRange {
+	return a.expressionPosition
+}
+
+func (a *AndBinaryOperation) Close() {
+	a.Left.Close()
+	a.Right.Close()
+}
+
+type andGroup struct {
+	leftSeriesCount      int
+	lastRightSeriesIndex int
+	rightSamplePresence  []bool // FIXME: this would be a good candidate for a bitmap type
+}
+
+// AccumulateRightSeriesPresence records the presence of samples on the right-hand side.
+func (g *andGroup) AccumulateRightSeriesPresence(data types.InstantVectorSeriesData, memoryConsumptionTracker *limiting.MemoryConsumptionTracker, timeRange types.QueryTimeRange) error {
+	if g.rightSamplePresence == nil {
+		var err error
+		g.rightSamplePresence, err = types.BoolSlicePool.Get(timeRange.StepCount, memoryConsumptionTracker)
+
+		if err != nil {
+			return err
+		}
+
+		g.rightSamplePresence = g.rightSamplePresence[:timeRange.StepCount]
+	}
+
+	for _, p := range data.Floats {
+		g.rightSamplePresence[timeRange.PointIndex(p.T)] = true
+	}
+
+	for _, p := range data.Histograms {
+		g.rightSamplePresence[timeRange.PointIndex(p.T)] = true
+	}
+
+	return nil
+}
+
+// FilterLeftSeries returns leftData filtered based on samples seen for the right-hand side.
+func (g *andGroup) FilterLeftSeries(leftData types.InstantVectorSeriesData, memoryConsumptionTracker *limiting.MemoryConsumptionTracker, timeRange types.QueryTimeRange) (types.InstantVectorSeriesData, error) {
+	filteredData := types.InstantVectorSeriesData{}
+
+	for idx, p := range leftData.Floats {
+		if !g.rightSamplePresence[timeRange.PointIndex(p.T)] {
+			continue
+		}
+
+		if filteredData.Floats == nil {
+			// First time we've seen a float sample that we should return, create a slice for the return values.
+			// We can reduce the length of the slice based on how many samples are left to check in the original series.
+
+			var err error
+			filteredData.Floats, err = types.FPointSlicePool.Get(len(leftData.Floats)-idx, memoryConsumptionTracker)
```

Comment:
could we re-use the leftData.Floats slice here somehow?

Reply:
We can, but it would come at the cost of not being able to possibly return a smaller capacity slice if the start of the series is filtered away. I suspect your suggestion will make things faster at the cost of slightly more complexity (which I'm OK with) - I'll give it a go and see what the benchmarks say.

Reply:
Done in 9298d92 - makes some benchmarks 1-2% faster 🎉